IBM HPCS/Key Protect KMS integration

humblec opened the pull request:
rbd: Implement Key Protect KMS integration for Ceph CSI
    
This commit adds the support for HPCS/Key Protect IBM KMS service
to Ceph CSI service. EncryptDEK() and DecryptDEK() of RBD volumes are
done with the help of key protect KMS server by wrapping and unwrapping
the DEK and by using the DEKStoreMetadata.
    

Fixes: # https://github.com/ceph/ceph-csi/issues/2585

What is covered: 
- [x] KMS integration interface
- [x] Documentation about the design consideration
- [x] Dependency update
- [x] Additionally this has been tested against Key Protect service hosted in an IBM cloud and below is the integration test result logs:
```
   

--------------- Application POD Mounting ------------
...
/pvc/name":"testkp","csi.storage.k8s.io/pvc/namespace":"openshift-storage","encrypted":"true","encryptionKMSID":"kp-metadata-test","imageFeatures":"layering","imageName":"csi-vol-948fe374-5f0d-11ec-a42d-0a580a84001c","journalPool":"ocs-storagecluster-cephblockpool","pool":"ocs-storagecluster-cephblockpool","storage.kubernetes.io/csiProvisionerIdentity":"1639727227436-8081-openshift-storage.rbd.csi.ceph.com","thickProvision":"false"},"volume_id":"0001-0011-...
.......
I1217 07:59:58.080408  257478 crypto.go:199] ID: 6 Req-ID: 0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c Encrypting device "/dev/rbd0"	 with LUKS
I1217 08:00:04.518933  257478 crypto.go:258] ID: 6 Req-ID: 0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c "/dev/mapper/luks-rbd-0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c" is not an active LUKS device (an error (exit status 4) occurred while running cryptsetup args: [status luks-rbd-0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c]):
I1217 08:00:04.518963  257478 crypto.go:210] ID: 6 Req-ID: 0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c Opening device "/dev/rbd0" with LUKS on "luks-rbd-0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c"
..
....



----- Inside Application POD -------------
...
/dev/mapper/luks-rbd-0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c  977M  2.5M  958M   1% /usr/share/nginx/html
tmpfs                                                                                                    15G   20K   15G   1% /run/secrets/kubernetes.io/serviceaccount

# 


----------- Unmounting -----------
...

I1217 08:25:42.631344  257478 crypto.go:232] ID: 68 Req-ID: 0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c Closing LUKS device "luks-rbd-0001-0011-openshift-storage-0000000000000001-948fe374-5f0d-11ec-a42d-0a580a84001c"
.....


------------------------ RBD image metadata which holds the Blob - -------------------------


bash-4.4$ rbd image-meta list -p ocs-storagecluster-cephblockpool csi-vol-6ec9281a-5f07-11ec-8323-0a580a84000f
There are 0 metadata on this image.
bash-4.4$ rbd image-meta list -p ocs-storagecluster-cephblockpool csi-vol-948fe374-5f0d-11ec-a42d-0a580a84001c\ 
> ;
There are 3 metadata on this image:

Key                                 Value                                                                                                                                                                                                                                                                           
rbd.csi.ceph.com/dek                ZXlKamFYQm9aWEowWlhoMElqb2ljRU40V1ZNd09VSjJXakp1TXpKaGMxTlVka1E1ZVU1SGRERlFORFZTU0RkRE5sQjFjVmhFYTJ0MWN6MGlMQ0pwZGlJNklsVm9TRVF5ZW5oQ1pGQk9jRVl5WkM4aUxDSjJaWEp6YVc5dUlqb2lOQzR3TGpBaUxDSm9ZVzVrYkdVaU9pSmpOMkU1WVdFNU1TMDFZMkkxTFRRNFpHRXRZVGd5TVMxbE9EVmpNamRpT1Rsa09USWlmUT09
rbd.csi.ceph.com/encrypted          encrypted                                                                                                                                                                                                                                                                       
rbd.csi.ceph.com/thick-provisioned  false  
```

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: Implement Key Protect KMS integration for Ceph CSI

This commit adds the support for HPCS/Key Protect IBM KMS service
to Ceph CSI service. EncryptDEK() and DecryptDEK() of RBD volumes are
done with the help of key protect KMS server by wrapping and unwrapping
the DEK and by using the DEKStoreMetadata.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
doc: IBM key protect/HPCS design doc

This commit add the design considerations of IBM Key protect KMS
service to the Ceph CSI integration.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: add kms unit test for key protect server

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+security module (HSM). It is designed to enable a user to take control of their
+cloud data encryption keys and cloud hardware security models. To support this
+KMS integration in Ceph CSI and thus enable the HPCS users to make use of the
+same in RBD volume encrypted operations, below things are considered.
```
please remove RBD from here as this can be later used for cephfs also

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+KMS integration in Ceph CSI and thus enable the HPCS users to make use of the
+same in RBD volume encrypted operations, below things are considered.
+
+## Connection to IBM HPCS/key Protect service
```
`key` to `Key`

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
````diff
@@ -0,0 +1,67 @@
+Below parameters/values can be used to establish the connection to the HPCS
+service from the CSI driver and to make use of the encryption operations:
+
+```console
````
should be text instead of console?

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+Ex: c7a9aa91-5cb5-48da-a821-e85c27b99d92
+
+* KP_REGION = [region of the key protect service]
+Ex: us-south-2
```
do we have any link on how to retrieve these pieces of information? if yes good to add it.

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+## Volume Encrypt or Decrypt Operation
+
+The IBM Key protect server's `wrap` and `unwrap` functionalities will be used by
+the Ceph CSI driver to achieve encryption and decryption of RBD volumes. The DEK
```
remove reference to rbd volume?

Madhu-1 commented on examples/kms/vault/csi-kms-connection-details.yaml:
```diff
@@ -60,5 +60,12 @@ data:
     }
+  kp-metadata-test: |-
+    {
+      "KMS_PROVIDER": "kp-metadata",
```
IMO  `kp-metadata` is a not good identifier can it be `ibm-encryption`?

humblec replied:
ibm encryption has been done by many services, so `kp` stands better. thanks.

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+* KMS_SERVICE_NAME=[kms_service_name]
+A unique name for the key management service within the project.
+
+* SERVICE_INSTANCE_ID=[service_instance_id]
```
it should be `KP_SERVICE_INSTANCE_ID`?

Madhu-1 commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+* SERVICE_API_KEY=[service_api_key]
+Ex:  06x6DbTkVQ-qCRmq9cK-p9xOQpU2UwJMcdjnIDdr0g2R
+
+* CUSTOMER_ROOT_KEY=[customer_root_key]
```
`CUSTOMER_ROOT_KEY` should be `KP_CUSTOMER_ROOT_KEY`?

Madhu-1 commented on go.mod:
```diff
@@ -3,6 +3,7 @@ module github.com/ceph/ceph-csi
 go 1.17
 
 require (
+	github.com/IBM/keyprotect-go-client v0.5.1
```
why not a new version as 0.7.0 is available?

humblec replied:
done

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+	KeyProtectSecretNameKey = "KMS_SECRET_NAME"
+	KeyProtectRegionKey     = "KP_REGION"
+
+	keyProtectServiceInstanceID = "KP_SERVICE_INSTANCE_ID"
+	// The following options are part of the Kubernetes Secrets.
+	// #nosec:G101, no hardcoded secrets, only configuration keys.
+	KeyProtectServiceAPIKey   = "KP_SERVICE_API_KEY"
+	KeyProtectCustomerRootKey = "KP_CUSTOMER_ROOT_KEY"
+
+	KeyProtectSessionToken = "KP_SESSION_TOKEN"
+	KeyProtectCRK          = "KP_CRK_ARN"
```
this will be used only for KP, not for HPCS? can we have a generic name if this will be used for both?

humblec replied:
regardless we use hpcs/key protect, the API server is same which is `Key Protect`, so this is the generic name,

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+	Initializer: initKeyProtectKMS,
+})
+
+type KeyProtectKMS struct {
```
add a comment for all the exported vars,types etc

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+	serviceInstanceID string
+	region            string
+	sessionToken      string
+	client            *kp.Client
```
move client as first element  lets not mix different types inside a struct

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+	keyProtectServiceInstanceID = "KP_SERVICE_INSTANCE_ID"
+	// The following options are part of the Kubernetes Secrets.
+	// #nosec:G101, no hardcoded secrets, only configuration keys.
+	KeyProtectServiceAPIKey   = "KP_SERVICE_API_KEY"
+	KeyProtectCustomerRootKey = "KP_CUSTOMER_ROOT_KEY"
+
+	KeyProtectSessionToken = "KP_SESSION_TOKEN"
+	KeyProtectCRK          = "KP_CRK_ARN"
```
these vars and below struct need to be exported?

humblec replied:
hmm.. actually it need not be.. but ,  `KeyProtect` stands as KMS name too and better read, but I can unexport too.

Madhu-1 replied:
please un-export which all are not required

humblec replied:
done.

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+			kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
```
this should be a string of interfaces? why not string of string?

humblec replied:
it is the same model we followed for all other kms interfaces integration so this .

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	// List keys in the KeyProtect instance
+	_, err = client.GetKeys(context.Background(), 0, 0)
+	if err != nil {
+		return fmt.Errorf("unable to fetch keys from key protect")
+	}
```
why this is needed? if needed why err is not returned back?

humblec replied:
it make sure we have the rbac..etc required for operating with the KMS server. I have returned that error too already?

Madhu-1 replied:
I don't think it's a valid reason. the RBAC errors will be returned during create operation. please remove this one and you are not returning/wrapping the original error message

humblec replied:
can be removed too. done.

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+
+	dekByteSlice := []byte(plainDEK)
+
+	result, err := kms.client.Wrap(context.Background(), kms.customerRootKey, dekByteSlice, nil)
```
why context.Backgroud why not context.TODO?

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+
+	// base64 encode the encrypted DEK, so that storing it should not have
+	// issues
+	encryptedDEK :=
+		base64.StdEncoding.EncodeToString(result)
```
no need to create an extra variable just return it.

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,234 @@
+			err)
+	}
+
+	result, err := kms.client.Unwrap(context.Background(), kms.customerRootKey, ciphertextBlob, nil)
```
use context.TODO?

humblec commented:
@Madhu-1 thanks.. I have addressed the comments. ptal.. thanks :+1:

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,233 @@
+
+	dekByteSlice := []byte(plainDEK)
+
+	result, err := kms.client.Wrap(context.TODO(), kms.customerRootKey, dekByteSlice, nil)
```
where this key is getting deletet? I don't see code for it am I missing something?

humblec replied:
its done as part of the plugin interface madhu..

Madhu-1 commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+// EncryptDEK uses the KeyProtect KMS and the configured CRK to encrypt the DEK.
+func (kms *KeyProtectKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
```
you don't need volumeID for encyption/decryption?

humblec replied:
not required.

nixpanic commented on examples/kms/vault/kms-config.yaml:
```diff
@@ -90,7 +90,13 @@ data:
         "secretName": "storage-encryption-secret"
-      }
+      },
+    "kp-metadata-test": {
```
indention looks  strange?

humblec replied:
corrected.

nixpanic commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,226 @@
+})
+
+// KeyProtectKMS store the KMS connection information retrieved from the kms configmap
+type KeyProtectKMS struct {
```
`keyProtectKMS` can probably be lowercase?

humblec replied:
All the kms structs are exported atm, so thought of keeping  it as exported for now and mark a todo for revisitng all other KMS and its references ..etc.

nixpanic replied:
Where did you mark this TODO?

humblec replied:
https://github.com/ceph/ceph-csi/issues/2725 @nixpanic this covers it.

nixpanic commented on internal/kms/keyprotect.go:
```diff
@@ -0,0 +1,226 @@
+
+	dekByteSlice := []byte(plainDEK)
+
+	result, err := kms.client.Wrap(context.TODO(), kms.customerRootKey, dekByteSlice, nil)
```
As the key is stored in the metadata of the volume, you need to make sure that the same passphrases are not encrypted the same way. This can usually be done by adding some salt to the encryption function, use the `volumeID` as salt if possible.

humblec replied:
yeah, however while checking the possibility of uniqueness by the randomness available : ( its random passphrase at start + wrapping algorithm randomness in KMS )  I didnt  add  `salt`, but can be done too as we have an unused arg here. will add that too

humblec replied:
Done.

nixpanic commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+### Values provided in the connection secret
+
+Considering `SERVICE_API_KEY` and `CUSTOMER_ROOT_KEY` are sensitive information,
+those will be provided as a kubernetes secret to the CSI driver. The Ceph CSI
```
nit: Kubernetes (capital), Kubernetes object like Secret also with capital letters (Secret, ConfigMap, Pod, ...)

humblec replied:
Done

nixpanic commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+
+Considering `SERVICE_API_KEY` and `CUSTOMER_ROOT_KEY` are sensitive information,
+those will be provided as a kubernetes secret to the CSI driver. The Ceph CSI
+KMS plugin interface for the key protect will read the secret name from the kms
```
uppercase KMS everywhere please

humblec replied:
Done.

nixpanic commented on docs/design/proposals/encryped-with-keyprotect.md:
```diff
@@ -0,0 +1,67 @@
+https://github.com/IBM/keyprotect-go-client provide the client SDK to interact
+with the Key Protect server and perform key protect operations.
+
+## Additional Reference#
```
trailing `#`

humblec replied:
corrected,

nixpanic commented:
Please also run `make containerized-tests` before updating the PR, make sure the simple formatting/linting CI jobs succeed.

humblec committed:
rebase: IBM key protect integration module dependency update

This commit adds the Key protect client SDK for the Key Protect
KMS integration to the Ceph CSI driver.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> Please also run `make containerized-tests` before updating the PR, make sure the simple formatting/linting CI jobs succeed.

yeah. mostly I verify that, but recently , my system started to report out of space before completing the run, so I was skipping it for last few times and kept running only `static-check` :) .. but will do the cleanup of the system and run it :)

humblec commented:
@nixpanic comments are addressed.. Thanks.. PTAL.
